pkg/operator/ceph/cluster/cleanup.go

```diff
@@ -0,0 +1,184 @@
+		labels[CleanupAppName] = "true"
+		job := &batch.Job{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      jobName,
+				Namespace: cluster.Namespace,
+				Labels:    labels,
+			},
+			Spec: batch.JobSpec{
+				Template: podSpec,
+			},
+		}
+
+		if err := k8sutil.RunReplaceableJob(c.context.Clientset, job, true); err != nil {
+			logger.Errorf("failed to run cluster clean up job on node %q. %v", hostName, err)
+		}
+	}
+}
+
+func (c *ClusterController) cleanUpJobContainer(cluster *cephv1.CephCluster) v1.Container {
+	volumeMounts := []v1.VolumeMount{}
+	envVars := []v1.EnvVar{}
+	if cluster.Spec.CleanupPolicy.DeleteDataDirOnHosts != "" && cluster.Spec.DataDirHostPath != "" {
+		hostPathVolumeMount := v1.VolumeMount{Name: volumeName, MountPath: cluster.Spec.DataDirHostPath}
+		volumeMounts = append(volumeMounts, hostPathVolumeMount)
+		envVars = append(envVars, []v1.EnvVar{
+			{Name: dataDirHostPath, Value: cluster.Spec.DataDirHostPath},
+			{Name: namespaceDir, Value: cluster.Namespace},
+		}...)
+	}
+	container := v1.Container{
+		Name:            "host-cleanup",
+		Image:           c.rookImage,
+		SecurityContext: osd.PrivilegedContext(),
+		VolumeMounts:    volumeMounts,
+		Env:             envVars,
+		Args:            []string{"ceph", "clean"}}
+
+	return container
+}
+
+func (c *ClusterController) cleanUpJobTemplateSpec(cluster *cephv1.CephCluster) v1.PodTemplateSpec {
+	volumes := []v1.Volume{}
+	hostPathVolume := v1.Volume{Name: volumeName, VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: cluster.Spec.DataDirHostPath}}}
+	volumes = append(volumes, hostPathVolume)
+
+	podSpec := v1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: CleanupAppName,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				c.cleanUpJobContainer(cluster),
+			},
+			Volumes:       volumes,
+			RestartPolicy: v1.RestartPolicyOnFailure,
+		},
+	}
+
+	return podSpec
+}
+
+func (c *ClusterController) waitForCephDaemonCleanUp(cluster *cephv1.CephCluster, retryInterval time.Duration) error {
+	logger.Infof("waiting for all the ceph daemons to be cleaned up in the cluster %q", cluster.Namespace)
+	for {
+		cephHosts, err := c.getCephHosts(cluster.Namespace)
+		if err != nil {
+			return errors.Wrap(err, "failed to list ceph daemon nodes")
+		}
+
+		if len(cephHosts) == 0 {
+			logger.Info("all ceph daemons are cleaned up")
+			break
+		}
+
+		logger.Debugf("waiting for ceph daemons in cluster %q to be cleaned up. Retrying in %q",
+			cluster.Namespace, retryInterval.String())
+		<-time.After(retryInterval)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) getCephHosts(namespace string) ([]string, error) {
+	cephPodCount := map[string]int{}
+	cephAppNames := []string{mon.AppName, mgr.AppName, osd.AppName, object.AppName, mds.AppName, rbd.AppName}
+	nodeNameList := util.NewSet()
+	hostNameList := []string{}
+
+	// get all the node names where ceph daemons are running
```

Comment:
Is this the right strategy? What if a monitor used to be running on a node but was failed over, and now no ceph daemons are on that node? This cleanup will fail to clean up that node. If cleanup isn't reliable, then users will need to do cleanup themselves, and if so, then what is the point of doing cleanup at all?

Reply:
As an improvement, the operator could keep track of all the nodes that mons and osds have ever been running on. But I would think 99%+ of scenarios would already be covered since we're looking for nodes where any ceph daemon is currently running so we could hold off on that additional bookkeeping until we confirm if it's needed.